src/scores/probability/fss_impl.py

```diff
@@ -0,0 +1,528 @@
+"""
+This module contains methods related to the FSS score
+
+For an explanation of the FSS, and implementation considerations,
+see: [Fast calculation of the Fractions Skill Score][fss_ref]
+
+[fss_ref]:
+https://www.researchgate.net/publication/269222763_Fast_calculation_of_the_Fractions_Skill_Score
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import TYPE_CHECKING, Iterable, Optional, Tuple, Union
+
+import numpy as np
+import numpy.typing as npt
+import xarray as xr
+
+from scores import utils
+from scores.typing import XarrayLike
+
+# Note: soft keyword `type` only support from >=3.10
+# "tuple" of 64 bit floats
+f8x3 = np.dtype("f8, f8, f8")
+# Note: `TypeAlias` on variables for python <=3.9
+if TYPE_CHECKING:  # pragma: no cover
+    FssDecomposed = Union[np.ArrayLike, np.DtypeLike]
+else:  # pragma: no cover
+    FssDecomposed = npt.NDArray[f8x3]
+
+
+class DimensionError(ValueError):
+    """
+    Wrapper for specific coordinate/dimension mismatch errors
+    """
+
+    # better to be explicit because python doesn't have braces for scoping its
+    # control flow
+    pass  # pylint: disable=unnecessary-pass
+
+
+class FssComputeMethod(Enum):
+    """
+    Choice of compute backend for FSS.
+    - Standard
+        - NUMPY (default)
+    - Optimized
+        - NUMBA_NATIVE (unavailable)
+        - NUMBA_PARALLEL (unavailable)
+    - Experimental
+        - RUST_NATIVE (unavailable)
+        - RUST_OCL (unavailable)
+    """
+
+    NUMPY = 1  # (default)
+    NUMBA_NATIVE = 2
+    NUMBA_PARALLEL = 3
+    RUST_NATIVE = 4
+    RUST_OCL = 5
+
+
+def fss_2d(
+    fcst: xr.DataArray,
+    obs: xr.DataArray,
+    *,  # Force keywords arguments to be keyword-only
+    threshold: np.float64,
+    window: Tuple[int, int],
+    spatial_dims: Tuple[str, str],
+    zero_padding: bool = False,
+    reduce_dims: Optional[Iterable[str]] = None,
+    preserve_dims: Optional[Iterable[str]] = None,
+    compute_method: FssComputeMethod = FssComputeMethod.NUMPY,
```

Comment:
I think the public API shouldn't include the compute method. In my mind they would be separate fuctions instead in different namespaces. e.g. scores.categorical.fss --> numpy version, scores.fast.fss --> numba version. Also, this is sitting in probability right now, but it should probably be in categorical since it operates on events not on continuous values or probability values.

Reply:
Right, I misunderstood the modules.

Basically this takes a continuous input, applies a threshold to transform it to categorical and then associates a probabilistic score based on accumulated window population populations essentially like the final score is like a Brier score, Happy to take your guidance on where it sits though.

Reply:
The type hinting for the dims should possibly be Optional[FlexibleDimensionTypes]

Reply:
The compute method is probably fine to keep around if it's going to make the implementation much easier, I'm just thinking as I read.

Reply:
Also as a follow up re: compute method, if we have scores.fast.fss wouldn’t the categorical/probability etc. be obfuscated by having to call it from a different module tree? I would also imagine the “fast” portion may not have to reimplement the whole algorithm - just the parts that may need to go fast (depending on the backend) - e.g. the aggregation logic is trivially parallel with ufunc and might not need numba, numba only needs to care about the FSS computation for a single field (just as an example). I feel like it’s intuitive to have a central point to choose the backend compute method…

Reply:
Let's have a chat tomorrow as well about some of the options here. If the end result is leaving this as-is, that's probably okay, but I think there might be some more options to discuss first.

Reply:
Similar resolution proposed as: https://github.com/nci/scores/pull/266#discussion_r1596174639

Re: place for the actual score. The backend will sit in `scores/fast/fss/...`. The implementation for contiunous inputs (where the user provides the threshold operator) sits in `scores/continous/fss_impl.py`. The categorical implementation where the user pre-computes the binary field, will sit in `scores/categorical/binary_fss_impl.py`.